### ISSUE TITLE
[RFC] Resolve Runtime Annotations via Interfaces

### ISSUE THREAD

**Issue by jellelicht** [actoron/jadex#15]
Currently, reflection to retrieve Runtime annotations such as `Belief` and `RequiredServices` is done on a class and all its superclasses in the `MicroClassReader` and all its Agent/Capability-annotated superclasses in the `BDIClassReader`.

Would you be open to accepting a PR to extend this mechanism to also work via interfaces?
This would make it possible to have e.g.
```java
@RequiredServices({                                                                                                                    
    @RequiredService(name="context", type=IContextService.class, binding=@Binding(scope=ServiceScope.PLATFORM))                        
})                                                                                                                                     
public interface Foo {
    @Belief
    long getBaz();
   
    @Belief
    void setBaz(long value);
}
```
, and allow having pretty ‘clean’ agent implementations for the actual classes.
E.g.:
```java
class MyAgent extends BDIAgent implements Foo …
```

Thanks for the consideration

**Comment by lb-actoron** [comment]
Interesting suggestion, basically you want to remove all the annotations from the pojo agent and waive them into using the interface, is that the correct interpretation? I am not really sure for what use cases that is helpful but it should be not too difficult to implement. Could you shortly comment if I understood that right?

**Comment by jellelicht** [comment]
Yes, you understood that right; although not _instead_ of agent annotations, but rather when the agent has none of these annotations. Imagine we have the following agent:
`class MyAgent extends ParentAgent implements IFooService`

Ideally, `ProvidedServices` annotations are resolved on `MyAgent` and then `ParentAgent`; my proposal is to also extend the search to `IFooService`.

There is still a discussion to be had about the multiple inheritance problem (what if several implemented interfaces come with a specific annotation? Solutions include ‘first-come-first-serve’, last-found, some kind of merging).

**Comment by lb-actoron** [comment]
Sure, I understand want you want to do but I was wondering why to have the annotations in an interface. Has the interface another purpose than providing the annotations, i.e. has the interface users (other objects that invoke methods on it)?

Multiple inheritance could be tackled by setting a resolution strategy; maybe requiring the user/programmer to explicitly set one and otherwise fail on incompatible annotations.

**Comment by jellelicht** [comment]
It allows for more flexible creation of objects that can be used in Jadex, and allows for more choices w.r.t. software architecture.

W.r.t. resolution strategies; the Spring folks have got this all figured out already with their AnnotationUtils, but I would not recommend adding a dependency on Spring.
